src/Streamly/Data/Serialize/Instances.hs

```diff
@@ -84,16 +87,33 @@ instance Serialize Aeson.Value where
 -- Vector.Vector
 --------------------------------------------------------------------------------
 
--- TODO: Serialize it independently
-
--- XXX Extremely inefficient serialization of Vector
 instance Serialize a => Serialize (Vector.Vector a) where
-    {-# INLINE size #-}
-    size i val = size i (Vector.toList val)
 
-    {-# INLINE deserialize #-}
-    deserialize off arr end =
-        fmap Vector.fromList <$> deserialize off arr end
+    {-# INLINE size #-}
+    size :: Int -> (Vector.Vector a) -> Int
+    size acc = foldl' size (acc + Unbox.sizeOf (Proxy :: Proxy Int))
 
     {-# INLINE serialize #-}
-    serialize off arr val = serialize off arr (Vector.toList val)
+    serialize :: Int -> MutableByteArray -> (Vector.Vector a) -> IO Int
+    serialize off arr val = do
+        let len = length val
+        finalOffset <- Vector.foldM (\curOff v -> serialize curOff arr v) (off+ Unbox.sizeOf (Proxy :: Proxy Int)) val
```

Comment:
- Indent this accordingly. 80 is the line limit.
- Use `Int64` over `Int`
- Add a space before `+` (`off + Unbox.sizeOf ...`)
- Can use `foldM'`